coded_tools/deep_rag/create_networks.py

```diff
@@ -0,0 +1,360 @@
+        Strongly consider overriding this method instead of the "easier" synchronous
+        invoke() version above when the possibility of making any kind of call that could block
+        (like sleep() or a socket read/write out to a web service) is within the
+        scope of your CodedTool and can be done asynchronously, especially within
+        the context of your CodedTool running within a server.
+
+        If you find your CodedTools can't help but synchronously block,
+        strongly consider looking into using the asyncio.to_thread() function
+        to not block the EventLoop for other requests.
+        See: https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread
+        Example:
+            async def async_invoke(self, args: Dict[str, Any], sly_data: Dict[str, Any]) -> Any:
+                return await asyncio.to_thread(self.invoke, args, sly_data)
+
+        :param args: An argument dictionary whose keys are the parameters
+                to the coded tool and whose values are the values passed for them
+                by the calling agent.  This dictionary is to be treated as read-only.
+        :param sly_data: A dictionary whose keys are defined by the agent hierarchy,
+                but whose values are meant to be kept out of the chat stream.
+
+                This dictionary is largely to be treated as read-only.
+                It is possible to add key/value pairs to this dict that do not
+                yet exist as a bulletin board, as long as the responsibility
+                for which coded_tool publishes new entries is well understood
+                by the agent chain implementation and the coded_tool implementation
+                adding the data is not invoke()-ed more than once.
+        :return: A return value that goes into the chat stream.
+        """
+        # Get most args as members
+        empty: Dict[str, Any] = {}
+        self.grouping_json: Dict[str, Any] = args.get("grouping_json", empty)
+        self.files_directory: str = args.get("files_directory")
+
+        # Main assembly of all networks we will deploy
+        reservationist: Reservationist = args.get("reservationist")
+        deployments: Dict[Reservation, Dict[str, Any]] = await self.assemble_deployments(reservationist)
+
+        # Deploy the reservations with confirmation event
+        # If you don't really need to wait until the new agent(s) has been deployed
+        # then set confirmation=False, and don't bother about waiting for the Event.
+        deployed_event: Event = None
+        try:
+            async with reservationist:
+                deployed_event = await reservationist.deploy(deployments, confirmation=True)
+
+        except ValueError as exception:
+            # Report exceptions from below as errors here.
+            error: str = f"{exception}"
+            self.logger.error(error)
+            return error
```

Comment:
FWIW: This allows network validation errors to be propagated to the text stream of the caller.